pandas/src/inference.pyx

```diff
@@ -655,22 +731,60 @@ def maybe_convert_numeric(object[:] values, set na_values,
         val = values[i]
 
         if val.__hash__ is not None and val in na_values:
+            seen_null = True
+            if _check_uint64_nan(seen_uint, seen_null,
```

Comment:
:smile: - indeed.  Unfortunate but necessary.  I agree that something could be simplified, but given that it is very performant at this point and most likely could be applied to `maybe_convert_objects`, it would be best for a future PR.